### PR TITLE
Generic host sample

### PR DESCRIPTION
**Pull request by danielmarbach**

### CODE REVIEW

**Comment by danielmarbach** [comment]
@timbussmann this is the basic structure of the code for the generic host. I haven't started doing any text because I figured I'd rather first get your consent if you think this is moving in the right direction

File: samples/hosting/generic-host/Core_7/GenericHost/GenericHost.csproj [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk.Worker">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <RuntimeIdentifier>win-x64</RuntimeIdentifier>
```
Comment (timbussmann) [inline samples/hosting/generic-host/Core_7/GenericHost/GenericHost.csproj]: why is this needed?

File: samples/hosting/generic-host/Core_7/GenericHost/GenericHost.csproj [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    <LangVersion>7.1</LangVersion>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="3.*" />
```
Comment (timbussmann) [inline samples/hosting/generic-host/Core_7/GenericHost/GenericHost.csproj]: why is this reference needed? Hosting also has a transitive dependency on the MS.Extensions.DI package

File: samples/hosting/generic-host/Core_7/GenericHost/Program.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    #region generic-host-critical-error
+
+    private static async Task OnCriticalError(ICriticalErrorContext context)
+    {
+        var fatalMessage =
+            $"The following critical error was encountered:{Environment.NewLine}{context.Error}{Environment.NewLine}Process is shutting down. StackTrace: {Environment.NewLine}{context.Exception.StackTrace}";
+
+        EventLog.WriteEntry(".NET Runtime", fatalMessage, EventLogEntryType.Error);
+
+        if (Environment.UserInteractive)
```
Comment (timbussmann) [inline samples/hosting/generic-host/Core_7/GenericHost/Program.cs]: when running in VS, console apps don't close automatically anymore so this shouldn't be needed anymore? Not sure it's the same with FailFast though?

File: samples/hosting/generic-host/Core_7/GenericHost/GenericHost.csproj (hunk 1 above)
Comment (danielmarbach) [reply samples/hosting/generic-host/Core_7/GenericHost/GenericHost.csproj]: It is not strictly needed but shows self-contained deployment according to https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/windows-service?view=aspnetcore-3.1&tabs=visual-studio#self-contained-deployment-scd

File: samples/hosting/generic-host/Core_7/GenericHost/Program.cs (hunk 3 above)
Comment (danielmarbach) [reply samples/hosting/generic-host/Core_7/GenericHost/Program.cs]: Ok but what about if you run it with the console switch instead with the debugger attached?

File: samples/hosting/generic-host/Core_7/GenericHost/GenericHost.csproj (hunk 2 above)
Comment (danielmarbach) [reply samples/hosting/generic-host/Core_7/GenericHost/GenericHost.csproj]: Yeah it was auto added for some reason. Removed. Good catch

File: samples/hosting/generic-host/Core_7/GenericHost/Program.cs (hunk 3 above)
Comment (timbussmann) [reply samples/hosting/generic-host/Core_7/GenericHost/Program.cs]: shouldn't we then ask for user input before closing too instead of just delaying?

Is this a bit too much functionality for a sample anyway?

File: samples/hosting/generic-host/Core_7/GenericHost/GenericHost.csproj (hunk 1 above)
Comment (timbussmann) [reply samples/hosting/generic-host/Core_7/GenericHost/GenericHost.csproj]: does this sample need to be optimized for self-contained deployment?

File: samples/hosting/generic-host/Core_7/GenericHost/Program.cs (hunk 3 above)
Comment (danielmarbach) [reply samples/hosting/generic-host/Core_7/GenericHost/Program.cs]: It was mentioned in the guidance so I figured I add it. I'm not opinionated enough to go one way or the other

File: samples/hosting/generic-host/Core_7/GenericHost/GenericHost.csproj (hunk 1 above)
Comment (danielmarbach) [reply samples/hosting/generic-host/Core_7/GenericHost/GenericHost.csproj]: It was mentioned in the guidance so I figured I add it. I'm not opinionated enough to go one way or the other

File: samples/hosting/generic-host/sample.md [hunk 4]
```diff
@@ -45,7 +45,7 @@ Received message #{Number}
 
 snippet: generic-host-service-lifetime
 
-The snippet above shows how the host builder runs by default as Windows Service. If the sample is started with the debugger attached it uses the console lifetime instead. If the generic host should always use as a console instead of a Windows Service use
+The snippet above shows how the host builder runs by default as Windows Service. If the sample is started with the debugger attached, it uses the console lifetime instead. If the generic host should always use as a console instead of a Windows Service use
```
Comment (andreasohlund) [inline samples/hosting/generic-host/sample.md]:
```suggestion
The snippet above shows how the host builder runs by default as a Windows Service. If the sample is started with the debugger attached, it uses the console lifetime instead. To always use as a console lifetime use
```

File: samples/hosting/generic-host/sample.md [hunk 5]
```diff
@@ -55,18 +55,18 @@ snippet: generic-host-logging
 
 To enable integration with Microsoft.Extensions.Logging the [`NServiceBus.MicrosoftLogging.Hosting`](https://www.nuget.org/packages/NServiceBus.MicrosoftLogging.Hosting/) community package is used and configured in combination with the standard logging.
 
-Next up the builder configures the usage of NServiceBus, this is achieved with the [`NServiceBus.Extensions.Hosting`](/nservicebus/hosting/extensions-hosting.md) package, including the [critical error](/nservicebus/hosting/critical-errors.md) action that will shutdown the application or service in case of a critical error.
+Next up, the builder configures the usage of NServiceBus, and this is achieved with the [`NServiceBus.Extensions.Hosting`](/nservicebus/hosting/extensions-hosting.md) package, including the [critical error](/nservicebus/hosting/critical-errors.md) action that will shut down the application or service in case of a critical error.
```
Comment (andreasohlund) [inline samples/hosting/generic-host/sample.md]:
```suggestion
Next up, the builder configures NServiceBus, using the [`NServiceBus.Extensions.Hosting`](/nservicebus/hosting/extensions-hosting.md) package, including the [critical error](/nservicebus/hosting/critical-errors.md) action that will shut down the application or service in case of a critical error.
```

File: samples/hosting/generic-host/sample.md [hunk 6]
```diff
@@ -55,18 +55,18 @@ snippet: generic-host-logging
...
 
 snippet: generic-host-nservicebus
 
 The critical error action:
 
 snippet: generic-host-critical-error
 
-To simulate work a BackgroundService called `Worker` is registered as a hosted services:
+To simulate work, a BackgroundService called `Worker` is registered as a hosted services:
```
Comment (andreasohlund) [inline samples/hosting/generic-host/sample.md]:
```suggestion
To simulate work, a BackgroundService called `Worker` is registered as a hosted service:
```

File: samples/hosting/generic-host/Core_7/GenericHost/Program.cs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        #endregion
+
+        #region generic-host-nservicebus
+
+        builder.UseNServiceBus(ctx =>
+        {
+            var endpointConfiguration = new EndpointConfiguration("Samples.Hosting.GenericHost");
+
+            endpointConfiguration.UseTransport<LearningTransport>();
+            endpointConfiguration.UsePersistence<LearningPersistence>();
```
Comment (andreasohlund) [inline samples/hosting/generic-host/Core_7/GenericHost/Program.cs]: Nitpick: Persistence isn't needed

**Comment by danielmarbach** [comment]
Pulling in now
